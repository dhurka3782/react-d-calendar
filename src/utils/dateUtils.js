import { getISOWeek } from 'date-fns';

export const getDaysInMonth = (date, weekStartDay = 1, calendarType = 'gregorian', showFixedNumberOfWeeks = false, showNeighboringMonth = true) => {
  if (calendarType !== 'gregorian') {
    throw new Error(`Unsupported calendar type: ${calendarType}. Only 'gregorian' is supported.`);
  }

  const year = date.getFullYear();
  const month = date.getMonth();
  const days = [];

  const firstDayOfMonth = new Date(year, month, 1);
  const lastDayOfMonth = new Date(year, month + 1, 0);
  const firstDayIndex = (firstDayOfMonth.getDay() + 7 - weekStartDay) % 7;

  // Days from previous month
  if (showNeighboringMonth) {
    const prevMonthLastDay = new Date(year, month, 0);
    for (let i = firstDayIndex - 1; i >= 0; i--) {
      days.push({ date: new Date(year, month - 1, prevMonthLastDay.getDate() - i), isCurrentMonth: false });
    }
  }

  // Days of current month
  for (let d = new Date(firstDayOfMonth); d <= lastDayOfMonth; d.setDate(d.getDate() + 1)) {
    days.push({ date: new Date(d), isCurrentMonth: true });
  }

  // Days from next month
  if (showNeighboringMonth || showFixedNumberOfWeeks) {
    const totalDays = days.length;
    const remainingDays = showFixedNumberOfWeeks ? 42 - totalDays : (Math.ceil(totalDays / 7) * 7) - totalDays;
    for (let i = 1; i <= remainingDays; i++) {
      days.push({ date: new Date(year, month + 1, i), isCurrentMonth: false });
    }
  }

  return days;
};

export const getWeeksInMonth = (date, weekStartDay) => {
  const days = getDaysInMonth(date, weekStartDay);
  const weeks = [];
  for (let i = 0; i < days.length; i += 7) {
    const weekStart = days[i].date;
    weeks.push(getISOWeek(weekStart));
  }
  return weeks;
};

export const getMonthsInYear = (date, showNeighboringDecade = false) => {
  const year = date.getFullYear();
  const months = Array.from({ length: 12 }, (_, i) => new Date(year, i, 1));
  if (showNeighboringDecade) {
    months.unshift(new Date(year - 1, 11, 1));
    months.push(new Date(year + 1, 0, 1));
  }
  return months;
};

export const getDecadesInCentury = (date, showNeighboringCentury = false) => {
  const year = date.getFullYear();
  const decadeStart = Math.floor(year / 10) * 10;
  const decades = Array.from({ length: 10 }, (_, i) => new Date(decadeStart + i * 10, 0, 1));
  if (showNeighboringCentury) {
    decades.unshift(new Date(decadeStart - 10, 0, 1));
    decades.push(new Date(decadeStart + 100, 0, 1));
  }
  return decades;
};