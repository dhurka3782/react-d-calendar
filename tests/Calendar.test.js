import React from 'react';
import { render, screen, fireEvent, within } from '@testing-library/react';
import Calendar from '../src/components/Calendar';
import { getDaysInMonth } from '../src/utils/dateUtils';

describe('Calendar component', () => {
  // Test 1: Check if the correct month and year are rendered
  test('renders correct month and year', () => {
    render(<Calendar date={new Date(2023, 0, 1)} />);
    expect(screen.getByText(/January 2023/i)).toBeInTheDocument();
  });

  // Test 2: Verify the correct number of days is rendered
  test('renders correct number of days', () => {
    render(<Calendar date={new Date(2023, 0, 1)} />);
    const daysContainer = screen.getByTestId('calendar-days');
    const dayButtons = within(daysContainer).getAllByRole('button');
    expect(dayButtons.length).toBe(31); 
  });

  // Test 3: Ensure navigation buttons update the month
  test('navigation buttons work', () => {
    render(<Calendar date={new Date(2023, 0, 1)} />);
    expect(screen.getByText(/January 2023/i)).toBeInTheDocument();

    const nextButton = screen.getByRole('button', { name: 'Next Month' });
    fireEvent.click(nextButton);
    expect(screen.getByText(/February 2023/i)).toBeInTheDocument();

    const prevButton = screen.getByRole('button', { name: 'Previous Month' });
    fireEvent.click(prevButton);
    expect(screen.getByText(/January 2023/i)).toBeInTheDocument();
    fireEvent.click(prevButton);
    expect(screen.getByText(/December 2022/i)).toBeInTheDocument();
  });

// Test 4: Confirm onDateSelect is called with the correct date
  test('onDateSelect callback is called', () => {
    const onDateSelect = jest.fn();
    render(<Calendar date={new Date(2023, 0, 1)} onDateSelect={onDateSelect} />);
    const day15 = screen.getByText('15');
    const dayButton = day15.closest('button');
    fireEvent.click(dayButton);
    expect(onDateSelect).toHaveBeenCalledTimes(1);
    const calledWith = onDateSelect.mock.calls[0][0];
    expect(calledWith.getFullYear()).toBe(2023);
    expect(calledWith.getMonth()).toBe(0);
    expect(calledWith.getDate()).toBe(15);
  });

  // Test 5: Test custom rendering of days
  test('custom renderDay is used', () => {
    const customRenderDay = (day) => <span className="custom-day">{day.getDate()}</span>;
    render(<Calendar date={new Date(2023, 0, 1)} renderDay={customRenderDay} />);
    const daysContainer = screen.getByTestId('calendar-days');
    const dayButtons = within(daysContainer).getAllByRole('button');
    const firstDay = dayButtons[0];
    expect(firstDay.querySelector('.custom-day')).toBeInTheDocument();
  });
});

describe('dateUtils', () => {
  // Test 6: Verify getDaysInMonth for a 31-day month
  test('getDaysInMonth returns correct days for January 2023', () => {
    const days = getDaysInMonth(new Date(2023, 0, 1));
    expect(days.length).toBe(31);
    expect(days[0].getDate()).toBe(1);
    expect(days[30].getDate()).toBe(31);
  });

  // Test 7: Verify getDaysInMonth for a 28-day month
  test('getDaysInMonth returns correct days for February 2023', () => {
    const days = getDaysInMonth(new Date(2023, 1, 1));
    expect(days.length).toBe(28);
    expect(days[0].getDate()).toBe(1);
    expect(days[27].getDate()).toBe(28);
  });

  // Test 8: Verify getDaysInMonth for a leap year
  test('getDaysInMonth handles leap years', () => {
    const days = getDaysInMonth(new Date(2024, 1, 1)); 
    expect(days.length).toBe(29);
    expect(days[0].getDate()).toBe(1);
    expect(days[28].getDate()).toBe(29);
  });

  // Test 9: Verify weekday headers for a month
  test('renders weekday headers', () => {
  render(<Calendar date={new Date(2023, 0, 1)} />);
  expect(screen.getByText('Sun')).toBeInTheDocument();
  expect(screen.getByText('Mon')).toBeInTheDocument();
  expect(screen.getByText('Sat')).toBeInTheDocument();
});
});