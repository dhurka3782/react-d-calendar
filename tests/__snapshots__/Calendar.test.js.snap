// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Calendar component matches snapshot 1`] = `
<div>
  <div
    aria-label="Interactive Calendar"
    class="calendar   theme-light custom-theme"
    style="--background-color: #f7fafc; --text-color: #2d3748; --primary-color: #4b6cb7; --accent-color: #48bb78; --header-background: #f7fafc; --disabled-color: #a0aec0;"
    tabindex="0"
  >
    <div
      class="header"
    >
      <div
        class="header-month-section"
      >
        <div
          class="header-month-nav"
        >
          <button
            aria-label="Previous"
            class="nav-button"
          >
            <svg
              aria-hidden="true"
              class="lucide lucide-chevron-left"
              fill="none"
              height="18"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="18"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="m15 18-6-6 6-6"
              />
            </svg>
          </button>
          <span
            class="header-month-label"
          >
            January 2025
          </span>
          <button
            aria-label="Next"
            class="nav-button"
          >
            <svg
              aria-hidden="true"
              class="lucide lucide-chevron-right"
              fill="none"
              height="18"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="18"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="m9 18 6-6-6-6"
              />
            </svg>
          </button>
        </div>
      </div>
      <div
        class="header-year-section"
      >
        <div
          aria-expanded="false"
          aria-label="Select year for month view"
          class="year-select-wrapper"
          role="combobox"
          tabindex="0"
        >
          <span
            class="year-select-display"
          >
            2025
          </span>
          <svg
            aria-hidden="true"
            class="lucide lucide-chevron-down chevron-down"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="1.5"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m6 9 6 6 6-6"
            />
          </svg>
        </div>
      </div>
    </div>
    <div
      class="calendar-container"
    >
      <div
        class="calendar-container "
      >
        <div
          class="calendar-instance "
        >
          <div
            class="calendar-weekdays"
          >
            <div
              class="weekday  "
            >
              THU
            </div>
            <div
              class="weekday  "
            >
              FRI
            </div>
            <div
              class="weekday  "
            >
              SAT
            </div>
            <div
              class="weekday  "
            >
              SUN
            </div>
            <div
              class="weekday  "
            >
              MON
            </div>
            <div
              class="weekday saturday "
            >
              TUE
            </div>
            <div
              class="weekday  sunday"
            >
              WED
            </div>
          </div>
          <div
            class="calendar-days"
          >
            <button
              aria-label="Select December 30, 2024"
              class="calendar-day
                  
                  
                  
                  adjacent-month"
              tabindex="0"
            >
              <span
                class="day-number"
              >
                30
              </span>
            </button>
            <button
              aria-label="Select December 31, 2024"
              class="calendar-day
                  
                  
                  
                  adjacent-month"
              tabindex="0"
            >
              <span
                class="day-number"
              >
                31
              </span>
            </button>
            <button
              aria-label="Select January 1, 2025"
              class="calendar-day"
              tabindex="0"
            >
              <span
                class="day-number"
              >
                1
              </span>
            </button>
            <button
              aria-label="Select January 2, 2025"
              class="calendar-day"
              tabindex="0"
            >
              <span
                class="day-number"
              >
                2
              </span>
            </button>
            <button
              aria-label="Select January 3, 2025"
              class="calendar-day"
              tabindex="0"
            >
              <span
                class="day-number"
              >
                3
              </span>
            </button>
            <button
              aria-label="Select January 4, 2025"
              class="calendar-day
                  
                  saturday"
              tabindex="0"
            >
              <span
                class="day-number"
              >
                4
              </span>
            </button>
            <button
              aria-label="Select January 5, 2025"
              class="calendar-day
                  
                  
                  sunday"
              tabindex="0"
            >
              <span
                class="day-number"
              >
                5
              </span>
            </button>
            <button
              aria-label="Select January 6, 2025"
              class="calendar-day"
              tabindex="0"
            >
              <span
                class="day-number"
              >
                6
              </span>
            </button>
            <button
              aria-label="Select January 7, 2025"
              class="calendar-day"
              tabindex="0"
            >
              <span
                class="day-number"
              >
                7
              </span>
            </button>
            <button
              aria-label="Select January 8, 2025"
              class="calendar-day"
              tabindex="0"
            >
              <span
                class="day-number"
              >
                8
              </span>
            </button>
            <button
              aria-label="Select January 9, 2025"
              class="calendar-day"
              tabindex="0"
            >
              <span
                class="day-number"
              >
                9
              </span>
            </button>
            <button
              aria-label="Select January 10, 2025"
              class="calendar-day"
              tabindex="0"
            >
              <span
                class="day-number"
              >
                10
              </span>
            </button>
            <button
              aria-label="Select January 11, 2025"
              class="calendar-day
                  
                  saturday"
              tabindex="0"
            >
              <span
                class="day-number"
              >
                11
              </span>
            </button>
            <button
              aria-label="Select January 12, 2025"
              class="calendar-day
                  
                  
                  sunday"
              tabindex="0"
            >
              <span
                class="day-number"
              >
                12
              </span>
            </button>
            <button
              aria-label="Select January 13, 2025"
              class="calendar-day"
              tabindex="0"
            >
              <span
                class="day-number"
              >
                13
              </span>
            </button>
            <button
              aria-label="Select January 14, 2025"
              class="calendar-day"
              tabindex="0"
            >
              <span
                class="day-number"
              >
                14
              </span>
            </button>
            <button
              aria-label="Select January 15, 2025"
              class="calendar-day"
              tabindex="0"
            >
              <span
                class="day-number"
              >
                15
              </span>
            </button>
            <button
              aria-label="Select January 16, 2025"
              class="calendar-day"
              tabindex="0"
            >
              <span
                class="day-number"
              >
                16
              </span>
            </button>
            <button
              aria-label="Select January 17, 2025"
              class="calendar-day"
              tabindex="0"
            >
              <span
                class="day-number"
              >
                17
              </span>
            </button>
            <button
              aria-label="Select January 18, 2025"
              class="calendar-day
                  
                  saturday"
              tabindex="0"
            >
              <span
                class="day-number"
              >
                18
              </span>
            </button>
            <button
              aria-label="Select January 19, 2025"
              class="calendar-day
                  
                  
                  sunday"
              tabindex="0"
            >
              <span
                class="day-number"
              >
                19
              </span>
            </button>
            <button
              aria-label="Select January 20, 2025"
              class="calendar-day"
              tabindex="0"
            >
              <span
                class="day-number"
              >
                20
              </span>
            </button>
            <button
              aria-label="Select January 21, 2025"
              class="calendar-day"
              tabindex="0"
            >
              <span
                class="day-number"
              >
                21
              </span>
            </button>
            <button
              aria-label="Select January 22, 2025"
              class="calendar-day"
              tabindex="0"
            >
              <span
                class="day-number"
              >
                22
              </span>
            </button>
            <button
              aria-label="Select January 23, 2025"
              class="calendar-day"
              tabindex="0"
            >
              <span
                class="day-number"
              >
                23
              </span>
            </button>
            <button
              aria-label="Select January 24, 2025"
              class="calendar-day"
              tabindex="0"
            >
              <span
                class="day-number"
              >
                24
              </span>
            </button>
            <button
              aria-label="Select January 25, 2025"
              class="calendar-day
                  
                  saturday"
              tabindex="0"
            >
              <span
                class="day-number"
              >
                25
              </span>
            </button>
            <button
              aria-label="Select January 26, 2025"
              class="calendar-day
                  
                  
                  sunday"
              tabindex="0"
            >
              <span
                class="day-number"
              >
                26
              </span>
            </button>
            <button
              aria-label="Select January 27, 2025"
              class="calendar-day"
              tabindex="0"
            >
              <span
                class="day-number"
              >
                27
              </span>
            </button>
            <button
              aria-label="Select January 28, 2025"
              class="calendar-day"
              tabindex="0"
            >
              <span
                class="day-number"
              >
                28
              </span>
            </button>
            <button
              aria-label="Select January 29, 2025"
              class="calendar-day"
              tabindex="0"
            >
              <span
                class="day-number"
              >
                29
              </span>
            </button>
            <button
              aria-label="Select January 30, 2025"
              class="calendar-day"
              tabindex="0"
            >
              <span
                class="day-number"
              >
                30
              </span>
            </button>
            <button
              aria-label="Select January 31, 2025"
              class="calendar-day"
              tabindex="0"
            >
              <span
                class="day-number"
              >
                31
              </span>
            </button>
            <button
              aria-label="Select February 1, 2025"
              class="calendar-day
                  
                  saturday
                  
                  adjacent-month"
              tabindex="0"
            >
              <span
                class="day-number"
              >
                1
              </span>
            </button>
            <button
              aria-label="Select February 2, 2025"
              class="calendar-day
                  
                  
                  sunday
                  adjacent-month"
              tabindex="0"
            >
              <span
                class="day-number"
              >
                2
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
